packages/devtools_app/test/shared/filter_test.dart

```diff
@@ -0,0 +1,295 @@
+      expect(controller.activeFilterTag(), 'ba cat:foobar');
+
+      // Only toggle filter active and no query filters.
+      controller.toggleFilters[0].enabled.value = true;
+      controller.setActiveFilter();
+      expect(controller.activeFilterTag(), equals('Hide multiples of 2'));
+
+      controller.toggleFilters[1].enabled.value = true;
+      controller.setActiveFilter();
+      expect(
+        controller.activeFilterTag(),
+        equals('Hide multiples of 2,Hide multiples of 3'),
+      );
+
+      controller.toggleFilters[0].enabled.value = false;
+      controller.setActiveFilter();
+      expect(controller.activeFilterTag(), equals('Hide multiples of 3'));
+
+      // Both query filter and toggle filters active.
+      controller.toggleFilters[0].enabled.value = true;
+      controller.toggleFilters[1].enabled.value = true;
+      controller.setActiveFilter(query: 'Ba cat:foobar');
+      expect(
+        controller.activeFilterTag(),
+        equals(
+          'Hide multiples of 2,Hide multiples of 3-#-ba cat:foobar',
+        ),
+      );
+    });
+
+    test('resetFilter', () {
+      // Verify default state.
+      controller.resetFilter();
+      _verifyBaseFilterState();
+
+      controller.toggleFilters[0].enabled.value = true;
+      controller.toggleFilters[1].enabled.value = true;
+      controller.setActiveFilter(query: 'cat:foobar');
+      for (final toggleFilter in controller.toggleFilters) {
+        expect(toggleFilter.enabled.value, isTrue);
+      }
+      expect(controller.activeFilter.value.queryFilter.isEmpty, isFalse);
+
+      controller.resetFilter();
+      _verifyBaseFilterState();
+    });
+  });
+}
+
+class _TestController extends DisposableController
+    with FilterControllerMixin<_TestDataClass>, AutoDisposeControllerMixin {
+  _TestController(this.data) {
+    subscribeToFilterChanges();
+  }
+
+  final List<_TestDataClass> data;
+
+  // Convenience getters for testing.
+  List<ToggleFilter<_TestDataClass>> get toggleFilters =>
+      activeFilter.value.toggleFilters;
+
+  Map<String, QueryFilterArgument> get queryFilterArgs =>
+      activeFilter.value.queryFilter.filterArguments;
+
+  @override
+  List<ToggleFilter<_TestDataClass>> createToggleFilters() => [
+        ToggleFilter<_TestDataClass>(
+          name: 'Hide multiples of 2',
+          includeCallback: (data) => data.id % 2 != 0,
+          enabledByDefault: true,
+        ),
+        ToggleFilter<_TestDataClass>(
+          name: 'Hide multiples of 3',
+          includeCallback: (data) => data.id % 3 != 0,
+        ),
+      ];
+
+  static const categoryFilterId = 'category-filter';
+
+  @override
+  Map<String, QueryFilterArgument> createQueryFilterArgs() => {
+        categoryFilterId: QueryFilterArgument(keys: ['cat', 'c']),
+      };
+
+  @override
+  void filterData(Filter<_TestDataClass> filter) {
+    super.filterData(filter);
+    if (filter.isEmpty) {
+      filteredData
+        ..clear()
+        ..addAll(data);
```

Comment:
We can remove the `else` if we just return here, which will make this a bit easier to read.